network/adaptive.go

```diff
@@ -0,0 +1,178 @@
+func NewCapabilities(changeC chan<- capability) *Capabilities {
+	return &Capabilities{
+		changeC: changeC,
+	}
+}
+
+// String implements Stringer interface
+func (c Capabilities) String() string {
+	var caps []string
+	for _, cap := range c.Flags {
+		caps = append(caps, fmt.Sprintf("%02x:%v", cap[0], pot.ToBin(cap[2:])))
+	}
+	return strings.Join(caps, ",")
+}
+
+// close the internal notification channel if it exists
+func (m Capabilities) destroy() {
+	if m.changeC != nil {
+		close(m.changeC)
+	}
+}
+
+// send a notification on the internal notification channel
+func (m Capabilities) notify(c capability) {
+	if m.changeC != nil {
+		m.changeC <- c
+	}
+}
+
+// get a capability bitvector for the module
+// the first two bytes returned are id and length, the following bytes contain the flags
+// returns nil if id is not registered
+func (c Capabilities) get(id uint8) capability {
+	if len(c.Flags) == 0 {
+		return nil
+	}
+	for _, caps := range c.Flags {
+		if caps[0] == id {
+			return caps
+		}
+	}
+	return nil
+}
+
+// adds a capability module to the bitvector collection
+// does not protect against duplicate ids. Calling code should use registerModule instead
+func (c *Capabilities) add(cap capability) {
+	c.Flags = append(c.Flags, cap)
+}
+
+// sets bits on bitvector
+// fails if:
+// * argument is 0-length byte
+// * capability module id is unknown
+// * the underlying set fails (see capability.set)
+func (c *Capabilities) set(id uint8, flags []byte) error {
+	if len(flags) == 0 {
+		return errors.New("flag bytes cannot be empty")
+	}
+	cap := c.get(id)
+	if cap == nil {
+		return fmt.Errorf("capability id %d not registered", id)
+	}
+	c.mu.Lock()
+	err := cap.set(flags)
+	ccopy := newCapability(cap[0], cap[1])
+	ccopy.set(cap[2:])
+	c.mu.Unlock()
+	if err != nil {
+		return err
+	}
+	c.notify(ccopy)
+	return nil
+}
+
+// unsets bits on bitvector
+// fails if:
+// * argument is 0-length byte
+// * capability module id is unknown
+// * the underlying set fails (see capability.set)
+func (c *Capabilities) unset(id uint8, flags []byte) error {
+	if len(flags) == 0 {
+		return errors.New("flag bytes cannot be empty")
+	}
+	cap := c.get(id)
+	if cap == nil {
+		return fmt.Errorf("capability id %d not registered", id)
+	}
+	c.mu.Lock()
+	err := cap.unset(flags)
+	ccopy := newCapability(cap[0], cap[1])
+	ccopy.set(cap[2:])
+	c.mu.Unlock()
+	if err != nil {
+		return err
+	}
+	c.notify(ccopy)
+	return nil
+}
+
+// adds a new module to the collection.
+// length is the length of the bitvector
+// fails if id already exists
+func (c *Capabilities) registerModule(id uint8, length uint8) error {
+	cap := c.get(id)
+	if cap != nil {
+		return fmt.Errorf("capability %d already registered", id)
+	}
+	cap = newCapability(id, length)
+	c.add(cap)
+	return nil
+}
+
+// capability is a single bitvector capabilities module
+type capability []byte
```

Comment:
We already have a `bitvector` type at `network/bitvector`. I think we can reuse it, rather than reimplement the data structure again here.

Reply:
Ah wasn't aware - thanks

Reply:
@nonsense using bitvector package made it quite a bit messier - and longer... :/

Reply:
@nolash it was a suggestion from my side, not a requirement. If you have a good reason not to use the `bitvector` and implement something new, then just say so, and do it :)

Reply:
Didnt know till I tried :) was worth trying, so thanks.